source/extensions/filters/network/thrift_proxy/decoder.h

```diff
@@ -0,0 +1,191 @@
+   * and the number of remaining container elements.
+   */
+  struct Frame {
+    Frame(ProtocolState state) : return_state_(state), elem_type_{}, value_type_{}, remaining_(0) {}
+    Frame(ProtocolState state, FieldType elem_type)
+        : return_state_(state), elem_type_(elem_type), value_type_{}, remaining_{} {}
+    Frame(ProtocolState state, FieldType elem_type, uint32_t remaining)
+        : return_state_(state), elem_type_(elem_type), value_type_{}, remaining_(remaining) {}
+    Frame(ProtocolState state, FieldType key_type, FieldType value_type, uint32_t remaining)
+        : return_state_(state), elem_type_(key_type), value_type_(value_type),
+          remaining_(remaining) {}
+
+    // Structs, lists, maps, and sets may be recursively nested in any combination. This field
+    // indicates which state to return to at the completion of each of those types.
+    const ProtocolState return_state_;
+
+    // Indicates the element type for lists and sets or the key type for a map.
+    const FieldType elem_type_;
+
+    // Indicates the value type for a map.
+    const FieldType value_type_;
+
+    // Indicates the number of elements (or key-value pairs) remaining in a list, map, or set.
+    uint32_t remaining_;
+  };
+
+  // These functions map directly to the matching ProtocolState values. Each returns the next state
+  // or ProtocolState::WaitForData if more data is required.
+  ProtocolState messageBegin();
+  ProtocolState messageEnd();
+  ProtocolState structBegin();
+  ProtocolState structEnd();
+  ProtocolState fieldBegin();
+  ProtocolState fieldValue();
+  ProtocolState fieldEnd();
+  ProtocolState listBegin();
+  ProtocolState listValue();
+  ProtocolState listEnd();
+  ProtocolState mapBegin();
+  ProtocolState mapKey();
+  ProtocolState mapValue();
+  ProtocolState mapEnd();
+  ProtocolState setBegin();
+  ProtocolState setValue();
+  ProtocolState setEnd();
+
+  // handleValue represents the generic Value state from the state machine documentation. It
+  // returns either ProtocolState::WaitForData if more data is required or the next state. For
+  // structs, lists, maps, or sets the return_state is pushed onto the stack and the next state is
+  // based on elem_type. For primitive value types, return_state is returned as the next state
+  // (unless WaitForData is returned).
+  ProtocolState handleValue(FieldType elem_type, ProtocolState return_state);
+
+  // handleState delegates to the appropriate method based on state_.
+  ProtocolState handleState();
+
+  // Helper method to retrieve the current frame's return state and remove the frame from the
+  // stack.
+  ProtocolState popReturnState();
+
+  Buffer::Instance& buffer_;
+  Protocol& proto_;
+  ProtocolState state_;
+  std::vector<Frame> stack_;
+};
+
+typedef std::unique_ptr<DecoderStateMachine> DecoderStateMachinePtr;
+
+/**
+ * Decoder encapsulates a configure TransportPtr and ProtocolPtr. It copies Thrift data into a
```

Comment:
That's fair. I think for the sniffing thrift filter copying may still need to take place. I have an idea for avoiding it there as well, but I'll have to see if it pans out.

The actual router implementation should be able to avoid copying when the downstream & upstream use the same transport and protocol. For that, we'll want the Decoder to use an externally managed buffer, so I'll just switch to that now.